packages/collections/persistent/map.pony

```diff
@@ -0,0 +1,557 @@
+    Return a new array with the provided node inserted at provided index,
+    shifting the old values at and above the index up one.
+    """
+    var belowArr: USize = 0
+    var aboveArr = idx
+    let newArray: Array[Map[K, V]] trn = recover trn Array[Map[K, V]] end
+    while(belowArr < idx) do
+      newArray.push(arr(belowArr))
+      belowArr = belowArr + 1
+    end
+    newArray.push(node)
+    while(aboveArr < arr.size()) do
+      newArray.push(arr(aboveArr))
+      aboveArr = aboveArr + 1
+    end
+    newArray
+
+  fun _overwriteInArrayAt(arr: Array[Map[K, V]] val,
+                          node: Map[K, V],
+                          idx: USize): Array[Map[K, V]] val ? =>
+    """
+    Return a new array with the provided node inserted at provided index.
+    """
+    var belowArr: USize = 0
+    var aboveArr = idx + 1
+    let newArray: Array[Map[K, V]] trn = recover trn Array[Map[K, V]] end
+    while(belowArr < idx) do
+      newArray.push(arr(belowArr))
+      belowArr = belowArr + 1
+    end
+    newArray.push(node)
+    while(aboveArr < arr.size()) do
+      newArray.push(arr(aboveArr))
+      aboveArr = aboveArr + 1
+    end
+    newArray
+
+  fun _removeInArrayAt(arr: Array[Map[K, V]] val,
+                       idx: USize): Array[Map[K, V]] val ? =>
+    """
+    Return a new array with the value at the provided index removed and values
+    above the index shifted down one.
+    """
+    var belowArr: USize = 0
+    var aboveArr = idx + 1
+    let newArray: Array[Map[K, V]] trn = recover trn Array[Map[K, V]] end
+    while(belowArr < idx) do
+      newArray.push(arr(belowArr))
+      belowArr = belowArr + 1
+    end
+    while(aboveArr < arr.size()) do
+      newArray.push(arr(aboveArr))
+      aboveArr = aboveArr + 1
+    end
+    newArray
+
+  fun remove(k: K): Map[K, V] ? =>
+    """
+    Try to remove the provided key from the node.
+    """
+    if contains(k) then
+      _removeWithHash(k, MapHelpers._hash[K](k), 0)
+    else
+      this as Map[K, V]
+    end
+
+  fun _removeWithHash(k: K, hash: U32, level: U32): Map[K, V] ? =>
+    """
+    Try to remove the prodied key form the node, using hash and level.
+    """
+    let bmapIdx = _BitOps.bitmapIdxFor(hash, level)
+    let arrayIdx: USize = _BitOps.arrayIdxFor(_bitmap, bmapIdx)
+    if level >= Maps._last_level() then
+      let newNode = _pointers(arrayIdx).remove(k)
+      MapNode[K, V](_bitmap, _overwriteInArrayAt(_pointers, newNode, arrayIdx))
+    else
+      let target = _pointers(arrayIdx)
+      if target._is_leaf() then
+        let newBMap = _BitOps.flipIndexedBitOff(_bitmap, bmapIdx)
+        let newArray = _removeInArrayAt(_pointers, arrayIdx)
+        MapNode[K, V](newBMap, newArray)
+      else
+        let newNode = target._removeWithHash(k, hash, level + 1)
+        let newArray = _overwriteInArrayAt(_pointers, newNode, arrayIdx)
+        MapNode[K, V](_bitmap, newArray)
+      end
+    end
+
+  fun getOrElse(k: K, alt: val->V): val->V =>
```

Comment:
Stylistically, I think the standard library is pretty consistent about using `snake_case` for method names, so we should probably stick with that in the stdlib packages going forward.